File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/KeyDerivationParams.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.dcbor.Cbor

/**
 * Sealed class wrapping the different key derivation parameter types.
 *
 * Each variant holds the parameters for a specific key derivation method.
 * This type is used internally by [EncryptedKey] to dispatch lock/unlock
 * operations to the appropriate implementation.
 */
sealed class KeyDerivationParams {

    /** HKDF-based key derivation. */
    data class HKDF(val params: HKDFParams) : KeyDerivationParams()

    /** PBKDF2-based key derivation. */
    data class PBKDF2(val params: PBKDF2Params) : KeyDerivationParams()

    /** Scrypt-based key derivation. */
    data class Scrypt(val params: ScryptParams) : KeyDerivationParams()

    /** Argon2id-based key derivation. */
    data class Argon2id(val params: Argon2idParams) : KeyDerivationParams()

    /** Returns the [KeyDerivationMethod] associated with these parameters. */
    fun method(): KeyDerivationMethod = when (this) {
        is HKDF -> KeyDerivationMethod.HKDF
        is PBKDF2 -> KeyDerivationMethod.PBKDF2
        is Scrypt -> KeyDerivationMethod.Scrypt
        is Argon2id -> KeyDerivationMethod.Argon2id
    }

    /** Returns `true` if the derivation method is password-based. */
    fun isPasswordBased(): Boolean = when (this) {
        is PBKDF2, is Scrypt, is Argon2id -> true
        else -> false
    }

    /**
     * Derives a key from [secret] and encrypts [contentKey] with it.
     *
     * The CBOR-encoded derivation parameters are stored in the AAD of the
     * returned [EncryptedMessage].
     */
    fun lock(contentKey: SymmetricKey, secret: ByteArray): EncryptedMessage = when (this) {
        is HKDF -> params.lock(contentKey, secret)
        is PBKDF2 -> params.lock(contentKey, secret)
        is Scrypt -> params.lock(contentKey, secret)
        is Argon2id -> params.lock(contentKey, secret)
    }

    /** Encodes these parameters as CBOR. */
    fun toCbor(): Cbor = when (this) {
        is HKDF -> params.toCbor()
        is PBKDF2 -> params.toCbor()
        is Scrypt -> params.toCbor()
        is Argon2id -> params.toCbor()
    }

    override fun toString(): String = when (this) {
        is HKDF -> params.toString()
        is PBKDF2 -> params.toString()
        is Scrypt -> params.toString()
        is Argon2id -> params.toString()
    }

    companion object {
        /**
         * Decodes [KeyDerivationParams] from a CBOR array.
         *
         * The first element of the array is the method index, which
         * determines the concrete parameter type.
         */
        fun fromCbor(cbor: Cbor): KeyDerivationParams {
            val a = cbor.tryArray()
            if (a.isEmpty()) {
                throw BcComponentsException.general("Empty KeyDerivationParams array")
            }
            val index = a[0].tryInt()
            val method = KeyDerivationMethod.fromIndex(index)
                ?: throw BcComponentsException.general("Invalid KeyDerivationMethod index: $index")
            return when (method) {
                KeyDerivationMethod.HKDF -> HKDF(HKDFParams.fromCbor(cbor))
                KeyDerivationMethod.PBKDF2 -> PBKDF2(PBKDF2Params.fromCbor(cbor))
                KeyDerivationMethod.Scrypt -> Scrypt(ScryptParams.fromCbor(cbor))
                KeyDerivationMethod.Argon2id -> Argon2id(Argon2idParams.fromCbor(cbor))
            }
        }
    }
}

File: csharp/BCComponents/BCComponents/KeyDerivationParams.cs
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// Wrapper for the different key derivation parameter types.
/// </summary>
/// <remarks>
/// Each variant holds the parameters for a specific key derivation method.
/// This type is used internally by <see cref="EncryptedKey"/> to dispatch
/// lock/unlock operations to the appropriate implementation.
/// </remarks>
public sealed class KeyDerivationParams : IEquatable<KeyDerivationParams>
{
    private enum Variant { HKDF, PBKDF2, Scrypt, Argon2id }

    private readonly Variant _variant;
    private readonly HKDFParams? _hkdf;
    private readonly PBKDF2Params? _pbkdf2;
    private readonly ScryptParams? _scrypt;
    private readonly Argon2idParams? _argon2id;

    private KeyDerivationParams(
        Variant variant,
        HKDFParams? hkdf,
        PBKDF2Params? pbkdf2,
        ScryptParams? scrypt,
        Argon2idParams? argon2id)
    {
        _variant = variant;
        _hkdf = hkdf;
        _pbkdf2 = pbkdf2;
        _scrypt = scrypt;
        _argon2id = argon2id;
    }

    /// <summary>Creates HKDF key derivation parameters.</summary>
    public static KeyDerivationParams FromHKDF(HKDFParams p) => new(Variant.HKDF, p, null, null, null);

    /// <summary>Creates PBKDF2 key derivation parameters.</summary>
    public static KeyDerivationParams FromPBKDF2(PBKDF2Params p) => new(Variant.PBKDF2, null, p, null, null);

    /// <summary>Creates Scrypt key derivation parameters.</summary>
    public static KeyDerivationParams FromScrypt(ScryptParams p) => new(Variant.Scrypt, null, null, p, null);

    /// <summary>Creates Argon2id key derivation parameters.</summary>
    public static KeyDerivationParams FromArgon2id(Argon2idParams p) => new(Variant.Argon2id, null, null, null, p);

    /// <summary>Returns the <see cref="KeyDerivationMethod"/> for these parameters.</summary>
    public KeyDerivationMethod Method => _variant switch
    {
        Variant.HKDF => KeyDerivationMethod.HKDF,
        Variant.PBKDF2 => KeyDerivationMethod.PBKDF2,
        Variant.Scrypt => KeyDerivationMethod.Scrypt,
        Variant.Argon2id => KeyDerivationMethod.Argon2id,
        _ => throw new InvalidOperationException(),
    };

    /// <summary>Returns <c>true</c> if the derivation method is password-based.</summary>
    public bool IsPasswordBased => _variant is Variant.PBKDF2 or Variant.Scrypt or Variant.Argon2id;

    /// <summary>Returns <c>true</c> if the derivation method is SSH agent-based.</summary>
    public bool IsSshAgent => false;

    /// <summary>
    /// Derives a key from <paramref name="secret"/> and encrypts
    /// <paramref name="contentKey"/> with it.
    /// </summary>
    public EncryptedMessage Lock(SymmetricKey contentKey, byte[] secret) => _variant switch
    {
        Variant.HKDF => _hkdf!.Lock(contentKey, secret),
        Variant.PBKDF2 => _pbkdf2!.Lock(contentKey, secret),
        Variant.Scrypt => _scrypt!.Lock(contentKey, secret),
        Variant.Argon2id => _argon2id!.Lock(contentKey, secret),
        _ => throw new InvalidOperationException(),
    };

    /// <summary>Encodes these parameters as CBOR.</summary>
    public Cbor ToCbor() => _variant switch
    {
        Variant.HKDF => _hkdf!.ToCbor(),
        Variant.PBKDF2 => _pbkdf2!.ToCbor(),
        Variant.Scrypt => _scrypt!.ToCbor(),
        Variant.Argon2id => _argon2id!.ToCbor(),
        _ => throw new InvalidOperationException(),
    };

    /// <summary>
    /// Decodes <see cref="KeyDerivationParams"/> from a CBOR array.
    /// </summary>
    /// <param name="cbor">The CBOR array value.</param>
    /// <returns>A new <see cref="KeyDerivationParams"/>.</returns>
    public static KeyDerivationParams FromCbor(Cbor cbor)
    {
        var a = cbor.TryIntoArray();
        if (a.Count == 0)
            throw BCComponentsException.General("Empty KeyDerivationParams array");

        var index = (int)a[0].TryIntoUInt64();
        var method = KeyDerivationMethodExtensions.FromIndex(index)
            ?? throw BCComponentsException.General($"Invalid KeyDerivationMethod index: {index}");

        return method switch
        {
            KeyDerivationMethod.HKDF => FromHKDF(HKDFParams.FromCbor(cbor)),
            KeyDerivationMethod.PBKDF2 => FromPBKDF2(PBKDF2Params.FromCbor(cbor)),
            KeyDerivationMethod.Scrypt => FromScrypt(ScryptParams.FromCbor(cbor)),
            KeyDerivationMethod.Argon2id => FromArgon2id(Argon2idParams.FromCbor(cbor)),
            _ => throw new InvalidOperationException(),
        };
    }

    /// <inheritdoc/>
    public override string ToString() => _variant switch
    {
        Variant.HKDF => _hkdf!.ToString(),
        Variant.PBKDF2 => _pbkdf2!.ToString(),
        Variant.Scrypt => _scrypt!.ToString(),
        Variant.Argon2id => _argon2id!.ToString(),
        _ => "Unknown",
    };

    /// <inheritdoc/>
    public bool Equals(KeyDerivationParams? other)
    {
        if (other is null) return false;
        if (_variant != other._variant) return false;
        return _variant switch
        {
            Variant.HKDF => _hkdf!.Equals(other._hkdf),
            Variant.PBKDF2 => _pbkdf2!.Equals(other._pbkdf2),
            Variant.Scrypt => _scrypt!.Equals(other._scrypt),
            Variant.Argon2id => _argon2id!.Equals(other._argon2id),
            _ => false,
        };
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is KeyDerivationParams p && Equals(p);

    /// <inheritdoc/>
    public override int GetHashCode() => _variant switch
    {
        Variant.HKDF => HashCode.Combine(_variant, _hkdf),
        Variant.PBKDF2 => HashCode.Combine(_variant, _pbkdf2),
        Variant.Scrypt => HashCode.Combine(_variant, _scrypt),
        Variant.Argon2id => HashCode.Combine(_variant, _argon2id),
        _ => 0,
    };
}
